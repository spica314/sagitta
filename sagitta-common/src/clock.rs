#[derive(Debug, Clone)]
pub struct Clock {
    pub fixed_time: Option<std::time::SystemTime>,
}

impl Clock {
    pub fn new() -> Self {
        Self { fixed_time: None }
    }

    pub fn new_with_fixed_time(fixed_time: std::time::SystemTime) -> Self {
        Self {
            fixed_time: Some(fixed_time),
        }
    }

    pub fn now(&self) -> std::time::SystemTime {
        self.fixed_time.unwrap_or_else(std::time::SystemTime::now)
    }

    pub fn is_fixed(&self) -> bool {
        self.fixed_time.is_some()
    }
}

impl Default for Clock {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_clock() {
        let clock = Clock::default();
        let now = clock.now();
        assert!(now.elapsed().unwrap().as_secs() < 1);
    }

    #[test]
    fn test_clock_with_fixed_time() {
        let fixed_time = std::time::SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(42);
        let clock = Clock::new_with_fixed_time(fixed_time);
        let now = clock.now();
        assert_eq!(now, fixed_time);
    }
}
