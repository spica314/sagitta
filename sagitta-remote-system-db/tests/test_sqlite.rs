use rand_pcg::Pcg64Mcg;
use sagitta_common::clock::Clock;
use sagitta_remote_system_db::{sqlite::SagittaRemoteSystemDbBySqlite, *};
use std::time::{Duration, SystemTime};
use tempfile::NamedTempFile;

#[test]
fn test_sqlite_workspace_1() {
    let file = NamedTempFile::new().unwrap();
    let path = file.into_temp_path();
    let path = path.to_path_buf();
    let clock = Clock::new_with_fixed_time(
        SystemTime::UNIX_EPOCH + Duration::from_secs(40 * 365 * 24 * 60 * 60),
    );
    let rng = Pcg64Mcg::new(42);
    let db = SagittaRemoteSystemDbBySqlite::new(path, rng, clock).unwrap();
    db.migration();

    let res1 = db
        .create_workspace(CreateWorkspaceRequest {
            workspace_name: "workspace1".to_string(),
        })
        .unwrap();
    insta::assert_debug_snapshot!(res1);
    let workspace1_id = res1.workspace_id;

    let res2 = db
        .get_workspaces(GetWorkspacesRequest {
            contains_deleted: false,
        })
        .unwrap();
    insta::assert_debug_snapshot!(res2);

    let res3 = db
        .delete_workspace(DeleteWorkspaceRequest {
            workspace_id: workspace1_id.clone(),
        })
        .unwrap();
    insta::assert_debug_snapshot!(res3);

    let res4 = db
        .get_workspaces(GetWorkspacesRequest {
            contains_deleted: false,
        })
        .unwrap();
    insta::assert_debug_snapshot!(res4);

    let res5 = db
        .get_workspaces(GetWorkspacesRequest {
            contains_deleted: true,
        })
        .unwrap();
    insta::assert_debug_snapshot!(res5);
}
