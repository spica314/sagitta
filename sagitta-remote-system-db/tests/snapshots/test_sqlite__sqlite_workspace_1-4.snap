---
source: sagitta-remote-system-db/tests/test_sqlite.rs
expression: res4
---
GetWorkspacesResponse {
    workspaces: [],
}
